engine/src/main/java/org/pentaho/di/trans/steps/orabulkloader/OraBulkLoader.java

```diff
@@ -0,0 +1,691 @@
+      } catch ( KettleFileException ex ) {
+        throw new KettleException( "Error retrieving badfile string", ex );
+      }
+    }
+
+    if ( meta.getDiscardFile() != null ) {
+      try {
+        FileObject fileObject =
+          getFileObject( meta.getDiscardFile(), getTransMeta() );
+
+        sb.append( " discard=\'" );
+        sb.append( getFilename( fileObject ) );
+        sb.append( "\'" );
+      } catch ( KettleFileException ex ) {
+        throw new KettleException( "Error retrieving discardfile string", ex );
+      }
+    }
+
+    DatabaseMeta dm = meta.getDatabaseMeta();
+    if ( dm != null ) {
+      String user = Const.NVL( dm.getUsername(), "" );
+      String pass =
+        Const.NVL( Encr.decryptPasswordOptionallyEncrypted( environmentSubstitute( dm.getPassword() ) ), "" );
+      if ( !password ) {
+        pass = "******";
+      }
+      String dns = Const.NVL( dm.getDatabaseName(), "" );
+      sb.append( " userid=" ).append( environmentSubstitute( user ) ).append( "/" ).append(
+        environmentSubstitute( pass ) ).append( "@" );
+
+      String overrideName = meta.getDbNameOverride();
+      if ( Utils.isEmpty( Const.rtrim( overrideName ) ) ) {
+        sb.append( environmentSubstitute( dns ) );
+      } else {
+        // if the database name override is filled in, do that one.
+        sb.append( environmentSubstitute( overrideName ) );
+      }
+    } else {
+      throw new KettleException( "No connection specified" );
+    }
+
+    if ( meta.isDirectPath() ) {
+      sb.append( " DIRECT=TRUE" );
+
+      if ( getStepMeta().getCopies() > 1 || meta.isParallel() ) {
+        sb.append( " PARALLEL=TRUE" );
+      }
+    }
+
+    return sb.toString();
+  }
+
+  public void checkExitVal( int exitVal ) throws KettleException {
+    if ( exitVal == EX_SUCC ) {
+      return;
+    }
+
+    if ( meta.isFailOnWarning() && ( exitVal == EX_WARN ) ) {
+      throw new KettleException( "sqlldr returned warning" );
+    } else if ( meta.isFailOnError() && ( exitVal != EX_WARN ) ) {
+      throw new KettleException( "sqlldr returned an error (exit code " + exitVal + ")" );
+    }
+  }
+
+  public boolean execute( OraBulkLoaderMeta meta, boolean wait ) throws KettleException {
+    Runtime rt = Runtime.getRuntime();
+
+    try {
+      sqlldrProcess = rt.exec( createCommandLine( meta, true ) );
+      // any error message?
+      StreamLogger errorLogger = new StreamLogger( sqlldrProcess.getErrorStream(), "ERROR" );
+
+      // any output?
+      StreamLogger outputLogger = new StreamLogger( sqlldrProcess.getInputStream(), "OUTPUT" );
+
+      // kick them off
+      errorLogger.start();
+      outputLogger.start();
+
+      if ( wait ) {
+        // any error???
+        int exitVal = sqlldrProcess.waitFor();
+        sqlldrProcess = null;
+        logBasic( BaseMessages.getString( PKG, "OraBulkLoader.Log.ExitValueSqlldr", "" + exitVal ) );
+        checkExitVal( exitVal );
+      }
+    } catch ( Exception ex ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)